### ISSUE TITLE
Crash caused by race condition in StatisticsViewModel

### ISSUE THREAD

**Issue by Rockefeller157** [AnnoDesigner/anno-designer#190]
Hello there!
Was using AD normally and then an error appeared. I post here a screenshot of the error and the log file created.

NB : I could continue to use it normally, until I clicked on "OK" on the error screen.

Logfile : https://pastebin.com/8uSiRuc2
Screenshot of the error : https://cdn.discordapp.com/attachments/326223813253529611/712033568074760273/unknown.png

**Comment by JakeYallop** [comment]
Hello, thanks for the crash report and the logs!

This is an issue we've ran into before - although this is the first time we've actually seen it crash the app.

A restart should fix the issue just fine - if you run into other issues just let us know.

If you don't mind, I'll hijack this issue to track the bug:).

**Comment by AgmasGold** [comment]
Hello, and thanks for crash report and the logs!

This is an issue we've ran into before, but its the first time we've actually seen it crash the app, and it doesn't seem to happen very often. A restart should fix the issue just fine.

If you don't mind, I'll hijack this issue report to track bug :).

**Comment by Rockefeller157** [comment]
Edit : It did that again.
Apparently it's related to roads as it happened a 2nd time after a restart, again by selecting and moving roads (x20 preset)

**Comment by AgmasGold** [comment]
Could we have a copy of the .ad file? It might be useful in debugging the issue.

**Comment by Rockefeller157** [comment]
https://cdn.discordapp.com/attachments/326223813253529611/712038624761217064/engineers-investors_cityp2.ad

There it is. Though I didn't save before it happened.

**Comment by AgmasGold** [comment]
Thanks - we'll try and figure out the exact issue.

**Comment by FroggieFrog** [comment]
I still can not reproduce the error, even while debugging. But the only potential solution I found is changing these lines

```csharp
var groupedBuildings = placedObjects.GroupBy(_ => _.Identifier);
var groupedSelectedBuildings = selectedObjects.Count > 0 ? selectedObjects.GroupBy(_ => _.Identifier) : null;

var buildingsTask = Task.Run(() => GetStatisticBuildings(groupedBuildings, buildingPresets));
var selectedBuildingsTask = Task.Run(() => GetStatisticBuildings(groupedSelectedBuildings, buildingPresets));
```

into this (by adding `.ToList()`)

```csharp
var groupedBuildings = placedObjects.GroupBy(_ => _.Identifier).ToList();
var groupedSelectedBuildings = selectedObjects.Count > 0 ? selectedObjects.GroupBy(_ => _.Identifier).ToList() : null;

var buildingsTask = Task.Run(() => GetStatisticBuildings(groupedBuildings, buildingPresets));
var selectedBuildingsTask = Task.Run(() => GetStatisticBuildings(groupedSelectedBuildings, buildingPresets));
```

@AgmasGold I know you also had the issue in debug mode. Does the proposed solution help?

**Comment by AgmasGold** [comment]
I'm not able to reproduce this reliably either. I have had it happen to me only a couple of times.

I'm not adding sure `.ToList()` will help - let me talk you through my reasoning, as I'm not 100% sure myself.

I think that the `selectedObjects` (or `placedObjects`) list is being modified on the main thread whilst enumeration is being done on the list in the statistics' threads. Calls to `GroupBy`, `Where` or any `IEnumerable` extension method creates a copy of the list, which is why we don't get any errors in the `GetStatisticBuildings` methods, as they are enumerating their own copy of the collections. This is why I don't think adding `.ToList()` will help.

 I think this error is caused when we run the `UpdateStatisticsAsync` method in quick succession, and something about the thread context gets updated. I'm not 100% certain, but I think the following might do a better job:

```csharp
lock (_statisticsLock)
{
    groupedBuildings = placedObjects.GroupBy(_ => _.Identifier);
    groupedSelectedBuildings = selectedObjects.Count > 0 ? selectedObjects.GroupBy(_    => _.Identifier) : null;
}
```

I'm still not sure if this would solve the issue though.

**Comment by FroggieFrog** [comment]
This should be fixed by PR #199.
If the issue is still present, feel free to reopen it. (should be included in a version >8.7)
